File: leetcode13.cpp
class Solution {
public:
	int getVal(char a)
	{
		switch(a)
		{
			case 'I':return 1;
			case 'V':return 5;
    		case 'X':return 10;
    		case 'L':return 50;
    		case 'C':return 100;
    		case 'D':return 500;
    		case 'M':return 1000;
		}
		return 0;
	}
    int romanToInt(string s) {
    	int res = 0;
    	char max = 'I';
    	for (int i = s.size()-1; i >= 0; --i)
    	{
    		if(getVal(s[i]) >= getVal(max))
    		{
    			max = s[i];
    			res = res + getVal(s[i]);
    		}
    		else
    			res = res - getVal(s[i]);
    	}
    	return res;
    }
};

File: leetcode13.java
public class Solution {
    private Map<Character, Integer> map = new HashMap<Character, Integer>(){{
        put('I',1);
        put('V',5);
        put('X',10);
        put('L',50);
        put('C',100);
        put('D',500);
        put('M',1000);
    }};
    
    public int romanToInt(String s) {
        int pre = 0, res = 0;
        for(char c : s.toCharArray()){
            int cur = map.get(c);
            res += (cur > pre) ? (cur - 2 * pre) : cur;
            pre = cur;
        }
        return res;
    }
}
